### PR TITLE
Data streams monitoring for Kafka

### PR DESCRIPTION
**Pull request by randomanderson**
# What Does this do

At a high level, this PR
- Adds an internal API for data streams monitoring and related reporting machinery
- Adds data streams feature detection
- Adds hooks to the Kafka client instrumentation for data streams monitoring
- Adds an implementation of FNV hashing

# Implementation Details
`PathwayContext` encapsulates the state around data streams propagation and checkpointing. It is added to the `DDSpanContext` and propagated to children in the same JVM. When DSM is disabled, a `NoopPathwayContext` is used. When DSM is enabled, `DefaultPathwayContext` is used. I opted to _always_ have a `PathwayContext` (rather than using null) because the locking concerns are much simpler. Earlier versions with `getOrCreate()` style methods were much more complicated.

Across queues (just Kafka in this PR), the `PathwayContext` is propagated with `encode`/`decode` and a message header. I created `BinaryKeyClassifier` and `BinaryContextVisitor` to mirror the existing inject/extract mechanism (`KeyClassifier`/`ContextVisitor`) for context injection. A simpler interface may be possible but I opted to follow what we already have.

Whenever a checkpoint is created, this generates a `StatsPoint` that is consumed by the `DefaultDataStreamsCheckpointer`. `StatsPoint`s are aggregated into `StatsGroup`s and reported every 10 seconds in `StatsBucket`s. The code of `DefaultDataStreamsCheckpointer` closely follows that of `ConflatingMetricsAggregator` since the logic for tracer metrics reporting is very similar.

The changes to queue instrumentations are fairly minimal. 3 lines of code were added to the kafka consumer instrumentation and only 1 line added to the producer instrumentation.

## Future work (in a future PR):
- Revisit shared code with tracer metrics. There's a lot of overlap
- Add instrumentation to get Kafka consumer group
- Health metrics
- Manual public API

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java [hunk 1]
```diff
@@ -104,6 +106,9 @@
 
   private final boolean disableSamplingMechanismValidation;
 
+  private volatile PathwayContext pathwayContext;
+  private final Random contextSelector = new Random();
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java:110]: Have you considered using `ThreadLocalRandom.current()` instead of creating a new random object per span context?

Comment (randomanderson) [reply dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java]: I made this change

File: internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/AgentSpan.java [hunk 2]
```diff
@@ -149,6 +149,8 @@ public interface AgentSpan extends MutableSpan, IGSpanInfo {
   /** RequestContext for the Instrumentation Gateway */
   RequestContext<Object> getRequestContext();
 
+  void mergePathwayContext(PathwayContext pathwayContext);
```
Comment (bantonsson) [inline internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/AgentSpan.java:152]: Really wish that we had abstracted away the attaching of a _context_ to the `AgentSpan`.

File: dd-trace-core/src/main/java/datadog/trace/core/datastreams/DefaultPathwayContext.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+  }
+
+  @Override
+  public void setCheckpoint(
+      String type, String group, String topic, Consumer<StatsPoint> pointConsumer) {
+
+    long startMillis = timeSource.getCurrentTimeMillis();
+    long nanoTime = timeSource.getNanoTicks();
+
+    lock.lock();
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/datastreams/DefaultPathwayContext.java:71]: How common is it that there will be races between calls to `setCheckpoint`?

File: internal-api/src/test/groovy/datadog/trace/util/FNV64HashTest.groovy [hunk 4]
```diff
@@ -0,0 +1,184 @@
+package datadog.trace.util
+
+import datadog.trace.test.util.DDSpecification
+import spock.lang.Requires
+
+// Test cases from http://www.isthe.com/chongo/src/fnv/test_fnv.c
+// This test requires Java8 for "Long.parseUnsignedLong(..)" although the underlying implementation does not
```
Comment (bantonsson) [inline internal-api/src/test/groovy/datadog/trace/util/FNV64HashTest.groovy:7]: You could use the parsing in `DDId`.

File: dd-trace-core/src/main/java/datadog/trace/core/CoreTracer.java [hunk 5]
```diff
@@ -716,6 +740,24 @@ public <C> void inject(AgentSpan span, C carrier, Setter<C> setter, PropagationS
     inject(span.context(), carrier, setter, style);
   }
 
+  @Override
+  public <C> void injectPathwayContext(
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/CoreTracer.java:744]: It feels like this could be part of the _regular_ context injection/extraction instead of being a separate mechanism, but looking at merging them can be for a separate PR.

**Review by bantonsson — APPROVED** [review]
I think this looks good, with some small comments and questions.

File: dd-trace-core/src/main/java/datadog/trace/core/datastreams/DefaultPathwayContext.java (hunk 3 above)
Comment (randomanderson) [reply dd-trace-core/src/main/java/datadog/trace/core/datastreams/DefaultPathwayContext.java:71]: It should be rare. New queue reads will generally create a new context, or if not, there's only one thread interacting anyway (eg, read followed by another read). In the second case, it's 50/50 whether the upstream context is used or the current context is reused (see `mergePipelineContext()`)

The race, if any, will probably between checkpointing and serialization/injection. Queue read followed by another read in the same thread, while also writing to another queue in another thread.

There will eventually be a public manual api so who know what kind of weird threading behavior people will decide to implement

Comment (bantonsson) [reply dd-trace-core/src/main/java/datadog/trace/core/datastreams/DefaultPathwayContext.java:71]: Thanks for the explanation. This use of `lock()` seems fine for now, and it's very self contained, so optimizing can be left for later.
